st2actions/st2actions/runners/ssh/paramiko_ssh.py

```diff
@@ -0,0 +1,406 @@
+        self.timeout = timeout
+        self.key_material = key_material
+        self.client = paramiko.SSHClient()
+        self.client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        self.logger = logging.getLogger(__name__)
+        self.sftp = None
+
+    def connect(self):
+        conninfo = {'hostname': self.hostname,
+                    'port': self.port,
+                    'username': self.username,
+                    'allow_agent': False,
+                    'look_for_keys': False}
+
+        if self.password:
+            conninfo['password'] = self.password
+
+        if self.key_files:
+            conninfo['key_filename'] = self.key_files
+
+        if self.key_material:
+            conninfo['pkey'] = self._get_pkey_object(key=self.key_material)
+
+        if not self.password and not (self.key_files or self.key_material):
+            conninfo['allow_agent'] = True
+            conninfo['look_for_keys'] = True
+
+        if self.timeout:
+            conninfo['timeout'] = self.timeout
+
+        extra = {'_hostname': self.hostname, '_port': self.port,
+                 '_username': self.username, '_timeout': self.timeout}
+        self.logger.debug('Connecting to server', extra=extra)
+
+        self.client.connect(**conninfo)
+        self.sftp = self.client.open_sftp()
+        return True
+
+    def put(self, local_path, remote_path, mode=None, mirror_local_mode=False):
+        extra = {'_local_path': local_path, '_remote_path': remote_path, '_mode': mode,
+                 '_mirror_local_mode': mirror_local_mode}
+        self.logger.debug('Uploading file', extra=extra)
+
+        if not os.path.exists(local_path):
+            raise Exception('Path %s does not exist locally.' % local_path)
+
+        rattrs = self.sftp.put(local_path, remote_path)
+
+        local_mode = mode
+        if not mode or mirror_local_mode:
+            local_mode = os.stat(local_path).st_mode
+
+        # Cast to octal integer in case of string
+        if isinstance(local_mode, basestring):
+            local_mode = int(local_mode, 8)
+        local_mode = local_mode & 07777
+        remote_mode = rattrs.st_mode
+        # Only bitshift if we actually got an remote_mode
+        if remote_mode is not None:
+            remote_mode = (remote_mode & 07777)
+        if local_mode != remote_mode:
+            self.sftp.chmod(remote_path, local_mode)
+
+        return rattrs
+
+    def put_dir(self, local_path, remote_path, mode=None, mirror_local_mode=False):
+        extra = {'_local_path': local_path, '_remote_path': remote_path, '_mode': mode,
+                 '_mirror_local_mode': mirror_local_mode}
+        self.logger.debug('Uploading dir', extra=extra)
+
+        if os.path.basename(local_path):
+            strip = os.path.dirname(local_path)
+        else:
+            strip = os.path.dirname(os.path.dirname(local_path))
+
+        remote_paths = []
+
+        for context, dirs, files in os.walk(local_path):
```

Comment:
Splitting this out into a utility functions and adding some tests for it would be nice :)